src/frontend/src/optimizer/rule/rewrite_like_expr_rule.rs

```diff
@@ -0,0 +1,160 @@
+// limitations under the License.
+
+use core::slice::memchr::memchr;
+use std::cmp::min;
+use std::str::from_utf8;
+
+use risingwave_common::types::ScalarImpl;
+use risingwave_connector::source::DataType;
+
+use super::{BoxedRule, Rule};
+use crate::expr::{Expr, ExprImpl, ExprRewriter, ExprType, ExprVisitor, FunctionCall, Literal};
+use crate::optimizer::plan_node::{ExprRewritable, LogicalFilter};
+use crate::optimizer::PlanRef;
+
+pub struct RewriteLikeExprRule {}
+impl Rule for RewriteLikeExprRule {
+    fn apply(&self, plan: PlanRef) -> Option<PlanRef> {
+        let filter: &LogicalFilter = plan.as_logical_filter()?;
+        let mut has_like = HasLikeExprVisitor {};
+        if filter
+            .predicate()
+            .conjunctions
+            .iter()
+            .any(|expr| has_like.visit_expr(expr))
+        {
+            let mut rewriter = LikeExprRewriter {};
+            Some(filter.rewrite_exprs(&mut rewriter))
+        } else {
+            None
+        }
+    }
+}
+
+struct HasLikeExprVisitor {}
+
+impl ExprVisitor<bool> for HasLikeExprVisitor {
+    fn merge(a: bool, b: bool) -> bool {
+        a | b
+    }
+
+    fn visit_function_call(&mut self, func_call: &FunctionCall) -> bool {
+        if func_call.get_expr_type() == ExprType::Like
+            && let (_, ExprImpl::InputRef(_), ExprImpl::Literal(_)) = func_call.clone().decompose_as_binary() {
+            true
+        } else {
+            func_call
+                .inputs()
+                .iter()
+                .map(|expr| self.visit_expr(expr))
+                .reduce(Self::merge)
+                .unwrap_or_default()
+        }
+    }
+}
+
+struct LikeExprRewriter {}
+
+impl ExprRewriter for LikeExprRewriter {
+    fn rewrite_function_call(&mut self, func_call: FunctionCall) -> ExprImpl {
+        let (func_type, inputs, ret) = func_call.clone().decompose();
+        let inputs: Vec<ExprImpl> = inputs
+            .into_iter()
+            .map(|expr| self.rewrite_expr(expr))
+            .collect();
+        if func_call.get_expr_type() == ExprType::Like
+            && let (_, ExprImpl::InputRef(x), ExprImpl::Literal(y)) = func_call.decompose_as_binary()
+            && matches!(y.return_type(), DataType::Varchar) {
+
+            let data = y.get_data();
+            if let Some(ScalarImpl::Utf8(data)) = data {
+                let bytes = data.as_bytes();
+                let len = bytes.len();
+                let idx = match (memchr(b'%', bytes), memchr(b'_', bytes)) {
+                    (Some(a), Some(b)) => min(a, b),
+                    (Some(idx), None) => idx,
+                    (None, Some(idx)) => idx,
+                    (None, None) => {
+                        return FunctionCall::new_unchecked(ExprType::Equal, inputs, ret).into()
+                    }
+                };
+
+                if idx == 0 {
+                    return FunctionCall::new_unchecked(ExprType::Equal, inputs, ret).into()
+                }
```

Comment:
Should it be:
```suggestion
                    (None, None) => {
                        return FunctionCall::new_unchecked(func_type, inputs, ret).into()
                    }
                };

                if idx == 0 {
                    return FunctionCall::new_unchecked(func_type, inputs, ret).into()
                }
```
? (or `func_call.into()` + clone)

It seems wrong

```
dev=> explain select * from t where x like '_';
                    QUERY PLAN                    
--------------------------------------------------
 BatchExchange { order: [], dist: Single }
 └─BatchFilter { predicate: (t.x = '_':Varchar) }
   └─BatchScan { table: t, columns: [x] }
(3 rows)
```